Switch to pip-compile-multi

peterdemin opened the pull request:
Hi, Zulip developers!

I'm a maintainer of [pip-compile-multi](https://pypi.python.org/pypi/pip-compile-multi/) - tool, that is meant to replace `tools/update-locked-requirements`.

In this PR I added `pip-compile-multi` to `requirements/dev.in` and executed:
```
$ pip-compile-multi
```

Some packages were updated and some stayed as is. I had to relax cryptography constraint because of version conflict.

It works a bit different from `tools/update-locked-requirements` - instead of copying all packages from referenced files, it inserts a reference to compiled requirements file. Other than that, it's just a wrapper for `pip-tools`.

(It's first battle test for the `pip-compile-multi` outside of my projects.)

rht commented:
Neat! Do you know any other projects that might benefit from `pip-compile-multi`?

I checked the CI build. There is an error in the production test https://travis-ci.org/zulip/zulip/jobs/329175801#L2832. Do you take into account that the package `future` and `futures` must be deleted manually in Python 3? This is a pip-tools bug. We patched this manually at 
 https://github.com/zulip/zulip/blob/374025e10bea776d2899cd826dedae25c761802f/tools/update-locked-requirements#L23-L27.

The error in the backend test is because you didn't bump the major version of `PROVISION_VERSION` at https://github.com/zulip/zulip/blob/master/version.py.

peterdemin commented:
Any Python service (by service I mean thing that is deployed by itself, as
opposed to libraries) should use version pinning, I use it in all projects
with my current employer.
I opened a PR in part to trigger test build. I don't have futures
workaround yet, but will surely add it.
And bump the version.
Once tests pass, how do you feel about switching to this tool?
On Tue, Jan 16, 2018 at 1:56 AM rht <notifications@github.com> wrote:

> Neat! Do you know any other projects that might benefit from
> pip-compile-multi?
>
> I checked the CI build. There is an error in the production test
> https://travis-ci.org/zulip/zulip/jobs/329175801#L2832. Do you take into
> account that the package future and futures must be deleted manually in
> Python 3? This is a pip-tools bug. We patched this manually at
>
> https://github.com/zulip/zulip/blob/374025e10bea776d2899cd826dedae25c761802f/tools/update-locked-requirements#L23-L27
> .
>
> The error in the backend test is because you didn't bump the major version
> of PROVISION_VERSION at
> https://github.com/zulip/zulip/blob/master/version.py.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/8079#issuecomment-357869747>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABYa2pm5vIFBFiKKcjLLtJQ76yMKskkVks5tLEgAgaJpZM4Rezxr>
> .
>

rht commented:
I am up for it, i.e. for the functionality to have a dedicated maintainer, and be reusable. Pinging @timabbott @showell.

Two last points:
1. How did you deal with the `-e` issues? `pip-tools` doesn't support url without `-e` prefix yet https://github.com/jazzband/pip-tools/issues/272.
2. You also need to check if the changes are consistent with the checks in `./tools/test-locked-requirements`. Making sure @adnrs96 sees this PR as well.

peterdemin commented:
1. The same way it was done in `zulip/tools/update-locked-requirements`: https://github.com/peterdemin/pip-compile-multi/blob/master/pipcompilemulti.py#L368-L376
2. Will do!

peterdemin commented:
I didn't realize that zulip has hard-pinned dependencies in both `*.in` and `*.txt` files. But `*.txt` having recursive dependencies too.
This way, recompiling lock files will generate the same result today and year after today. So it's possible to validate, that compilation was run after changes to *.in files were made.
However, this approach makes updating dependencies somewhat obscure...

I used to have minimal set of version constraints in `.in` file - only for incompatibilities (or when an upgrade will take more time, that is available). With more than 100 dependencies compilation will produce (slightly) different results almost every day. Which makes validation by recompilation useless.

What if instead of hashing results, pip-compile-multi will save hash for `.in` files and check, that hashes still match in CI build?

rht commented:
> With more than 100 dependencies compilation will produce (slightly) different results almost every day.

> What if instead of hashing results, pip-compile-multi will save hash for .in files and check, that hashes still match in CI build?

Actually `./tools/test-locked-requirements` runs the *.txt creation every single time the CI is triggered / whenever the backend test is run manually in local env. They result in the same *.txt output every single time if the *.in files are unchanged. This is to ensure deterministic build of the dependencies.

rht commented:
There is a conflict because of https://github.com/zulip/zulip/commit/167a71292283f37cdd821856c7db95cf82fea419.

adnrs96 commented:
yeah, validation by recompilation won't be useless since we will always have pinned versions of dependencies in the `*.in` files. So a change in hash of `*.in` will mean there is some change in the `*.txt` files and we will get the diff for it if someone didn't update the locked requirements using `tools/update-locked-requirements`.
Theory behind having pinned dependencies in `*.in` files (which are manually added by developers when needed) and then getting all the recursive dependencies in the `*.txt` files is that Zulip installs dependencies from files using the `--no-deps` option. This ensures that even if the dependency itself doesn't use pinned version of dependencies for its requirements file (and hence even might be at risk of breaking someday) Zulip still stays on the safe side. The `*.txt` files will have all the dependencies we will need to get with pinned versions.

I just have one doubt here. I don't know what is the behaviour of `pip` when it encounters a `-r common.txt` in `dev.txt` or whatever be the file names. Like currently we have all the deps in `dev.txt` that will be required for a successful install of development environment. Does `pip` automatically go to `common.txt` to get the rest of compiled deps? Especially when the `--no-deps` option is being used.

Apart from this, I think we will just need to add `common.txt` and similarly other new files which will now be generated (due to top down approach of the `pip-compile-multi`) to the `tools/test_locked_requirements` for it to work.

rht commented:
> I don't know what is the behaviour of pip when it encounters a -r common.txt in dev.txt or whatever be the file names.

IIRC, `-r common.txt` is used in dev.txt / prod.txt pre-pip-tools days and things were working fine as well.

So your interpretation of what is going on
> Does pip automatically go to common.txt to get the rest of compiled deps? Especially when the --no-deps option is being used.

is most likely correct.

adnrs96 commented:
then I guess we just need to do 
> Apart from this, I think we will just need to add common.txt and similarly other new files which will now be generated (due to top down approach of the pip-compile-multi) to the tools/test_locked_requirements for it to work.

and we should be fine. Everything else will be pretty much same, I think.

peterdemin commented:
I read the discussion three times, but didn't get, what needs to be done :)
In the meantime, I liked the idea of checking whether locked files need recompilation.
Please comment on https://github.com/peterdemin/pip-compile-multi/issues/5

rht commented:
The error https://travis-ci.org/zulip/zulip/jobs/332512424#L1311 means dev.txt or prod.txt is generated without the knowledge of the content of common.txt. Is it possible to generate them later?

peterdemin commented:
`pip-compile-multi` sorts requirements files topologically, so all dependencies are compiled before files, that include them.
The cause for this error is that babel is included in both `doc.txt` and `common.txt` with different versions. The solution is to relax constraint in `doc.in` (I just pushed it).

rht commented:
I see. It looks like `certifi` needs to be `>=2017.7.27.1` in `common.in`.

peterdemin commented:
Yeah, and after `certifi` same issue arised for `jinja2`. So I replaced all `==` with `~=` (except for `git+http` deps) in `common.in` and `docs.in`.
I'm not sure that it's the best way to go.  Having the same packages in two files (docs.in and common.in) will make more troubles in future for sure.
One solution is to reference common from docs - so `docs.txt` will have only additional packages needed to build docs.

peterdemin commented:
I have mixed feelings about this test failure: https://travis-ci.org/zulip/zulip/jobs/332840863#L3351
```
Jan 24 15:55:11 +++ /home/travis/success-http-headers.txt	2018-01-24 15:55:11.025760056 +0000
Jan 24 15:55:11 @@ -1,4 +1,3 @@
Jan 24 15:55:11 -  Self-signed certificate encountered.
Jan 24 15:55:11  WARNING: no certificate subject alternative name matches
Jan 24 15:55:11  	requested host name ‘localhost’.
Jan 24 15:55:11    HTTP/1.1 302 Found
Jan 24 15:55:11 
Jan 24 15:55:11 FAILURE: The HTTP Headers returned from loading the homepage on the server do not match the contents of tools/travis/success-http-headers.txt.  Typically, this means that the server threw a 500 when trying to load the homepage.
Jan 24 15:55:11 Displaying the contents of the server's error log:
Jan 24 15:55:11 
Jan 24 15:55:11 
Jan 24 15:55:11 Displaying the contents of the main server log:
Jan 24 15:55:11 
Jan 24 15:55:11 2018-01-24 15:55:09.236 INFO [] Tornado loaded 0 event queues in 0.000s
Jan 24 15:55:11 2018-01-24 15:55:09.785 INFO [zr] 127.0.0.1       GET     302  44ms (+start: 9.0s) / (unauth via ?)
Jan 24 15:55:11 2018-01-24 15:55:10.321 INFO [zr] 127.0.0.1       GET     301 534ms /login (unauth via ?)
Jan 24 15:55:11 2018-01-24 15:55:11.023 INFO [zr] 127.0.0.1       GET     200 160ms (db: 3ms/3q) (+start: 540ms) /login/ (unauth via ?)
```
Is it possible that the `302 Found` was caused by dependency update?

timabbott commented:
@peterdemin that success-http-headers error is unrelated to your changes; it was a bug briefly in master.  Rebase and it should go away.

peterdemin commented:
Now I got
```
django.db.utils.OperationalError: could not extend file "base/22216/20141": No space left on device
HINT:  Check free disk space.
```
Maybe if you retry the job, it will pass?

rht commented:
The CircleCI still outputs error. https://circleci.com/gh/zulip/zulip/672?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link.

The error is because you didn't replace this line https://github.com/zulip/zulip/blob/master/tools/test-locked-requirements#L46. (And additionally the error message in https://github.com/zulip/zulip/blob/master/tools/test-locked-requirements#L120).

Also, reminder that you haven't bumped the provision version as I stated in https://github.com/zulip/zulip/pull/8079#issuecomment-357869747.

rht commented:
> The error is because you didn't replace this line https://github.com/zulip/zulip/blob/master/tools/test-locked-requirements#L46. (And additionally the error message in https://github.com/zulip/zulip/blob/master/tools/test-locked-requirements#L120).

Ah ic nvm, I didn't see that `update-locked-requirements` becomes a wrapper to `pip-compile-multi`. Not sure why it still fails tho.

rht commented:
It looks like because it is using `--output-dir`.

peterdemin commented:
Yes, `pip-compile-multi` doesn't support --output-dir. Instead, I can copy whole requirements directory and run `pip-compile-multi` inside it.

peterdemin commented:
I wonder if `tools/update-locked-requirements` can be deleted at this point?
Also, what do you think, can [`pip-compile-multi verify`](https://github.com/peterdemin/pip-compile-multi#check-that-pip-compile-multi-was-run-after-changes-in-in-file) replace `tools/test-locked-requirements`? They use different mechanics, but I think the result is the same.

peterdemin commented:
Currently, CircleCI fails with
```
Jan 26 02:25:33 Running API tests...
Jan 26 02:25:33 API tests passed!
+ ./tools/test-locked-requirements
Traceback (most recent call last):
  File "./tools/test-locked-requirements", line 124, in <module>
    main()
  File "./tools/test-locked-requirements", line 104, in main
    requirements_are_consistent = test_locked_requirements()
  File "./tools/test-locked-requirements", line 47, in test_locked_requirements
    stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
  File "/usr/lib/python3.4/subprocess.py", line 561, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['pip-compile-multi', '--directory', '/var/tmp/zulip-requirements']' returned non-zero exit status 1
Exited with code 1
```
I assumed, that `pip-compile-multi` is installed at this point (because non-built-in `ujson` is imported) but maybe I was wrong?

rht commented:
Strange, https://travis-ci.org/zulip/zulip/jobs/333552272#L4857 said it passed just fine. Maybe `pip-compile-multi` is installed but not found in the path?

Also, I'm not sure why this is the case
```
    stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
```
Maybe we should see what the error message actually is to clear the fog.

peterdemin commented:
Great idea! I don't know why I left it there...

peterdemin commented:
From the logs I understood, that pip-compile fails to check `egg_info` for pycurl `7.43.0.1` (thumbor dependency).
I've set it back to `7.43.0`, hope that helps

peterdemin commented:
Argh, didn't help. I don't have Windows, can someone reproduce CircleCI error and check what's wrong with `pycurl`?

rht commented:
The CircleCI test runs on Ubuntu 14.04 (it is meant to replace the Travis CI at some point in the future).
What about setting pycurl ~= 7.43.0?

peterdemin commented:
Ah, I thought you're using Travis and CircleCI for different OSes. I'll debug the issue under CentOS and if I won't find anything, I'll try Ubuntu later.
Regarding pycurl version - it didn't help, I think, I will remove it back from `.in` file.

gnprice commented:
@peterdemin One tool that can be very handy in debugging -- CircleCI actually gives you the ability to SSH into the build! See https://circleci.com/docs/2.0/ssh-access-jobs/ . That's one of the reasons we're moving to it from Travis.

(I'm not sure if it'll give you that option directly on [the zulip/zulip build](https://circleci.com/gh/zulip/zulip/711) that the status box on this PR thread links to. If you don't see it when you go to that page, you can [set up CircleCI](https://circleci.com/signup/) for your own GitHub fork of the repo, and that should do it.)

peterdemin commented:
@gnprice Thanks for the info. Actually, I managed to reproduce the issue in Cloud9. I still don't know how to solve it or why it didn't raise before.
In short, `pip-compile` executes `python setup.py egg_info` to get the list of package dependencies. `pycurl` uses `distutils` instead of `setuptools`, which doesn't support this command:
```
$ python setup.py egg_info                                                                                                                     
Using curl-config (libcurl 7.53.1)
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

error: invalid command 'egg_info'
```
I tried to replace distutils import with setuptools and it solved the issue... I googled similar issues on StackOverflow, but all suggested fixes didn't work. I'm kinda stuck here. Do you have any ideas?

gnprice commented:
> Do you have any ideas?

I don't, unfortunately. Python packaging is pretty complicated, and I've always managed to avoid having to become an expert on it. Hopefully it's an informative example for your project!

rht commented:
> I tried to replace distutils import with setuptools and it solved the issue...

This should have been detected by pip-tools, but it would also suffice to detect the error and have pip-compile-multi handle the distutils case until the fix is delivered upstream?

peterdemin commented:
`pip-compile` fails to generate `.txt` file because of this error, so I can't think of a way to work around it.
AFAIK `python setup.py egg_info` is the (only?) best way of getting a list of package dependencies, and it fails for pycurl.

peterdemin commented:
I'll try to make a minimal working example and see why pip-compile handles it differently when running with and without `pip-compile-multi`

peterdemin commented:
As I understand, pip-tools released a workaround for this problem. So I rebased my changes to see if it will work on CircleCI.
Now it shows different error:
```
Feb 09 03:12:28 Traceback (most recent call last):
Feb 09 03:12:28   File "/srv/zulip-thumbor-venv/bin/thumbor", line 7, in <module>
Feb 09 03:12:28     from thumbor.server import main
Feb 09 03:12:28   File "/srv/zulip-venv-cache/b81ca633007b175aa46374073a802cb70a297637/zulip-thumbor-venv/local/lib/python2.7/site-packages/thumbor/server.py", line 22, in <module>
Feb 09 03:12:28     from thumbor.console import get_server_parameters
Feb 09 03:12:28   File "/srv/zulip-venv-cache/b81ca633007b175aa46374073a802cb70a297637/zulip-thumbor-venv/local/lib/python2.7/site-packages/thumbor/console.py", line 13, in <module>
Feb 09 03:12:28     from thumbor.context import ServerParameters
Feb 09 03:12:28   File "/srv/zulip-venv-cache/b81ca633007b175aa46374073a802cb70a297637/zulip-thumbor-venv/local/lib/python2.7/site-packages/thumbor/context.py", line 13, in <module>
Feb 09 03:12:28     from concurrent.futures import ThreadPoolExecutor, Future
```
For some reason it uses python2.7...

peterdemin commented:
I dug the tools and discovered, that thumbor runs under Python2.7 while everything else under Python3.x.
Since I compiled all requirements under Python3 thumbor.txt didn't get backport for concurrent.futures.
So, I recompiled thumbor.in under Python2.7 and got futures into thumbor.txt

peterdemin commented:
Alright, now I get
```
    raise Exception("It looks like you have updated some python dependencies but haven't "
Exception: It looks like you have updated some python dependencies but haven't updated locked requirements files. Please update them by running `tools/update-locked-requirements`. For more information please refer to `requirements/README.md`.
```
which must be easy to fix. But I need some sleep.

rht commented:
> I dug the tools and discovered, that thumbor runs under Python2.7 while everything else under Python3.x.
> Since I compiled all requirements under Python3 thumbor.txt didn't get backport for concurrent.futures. So, I recompiled thumbor.in under Python2.7 and got futures into thumbor.txt

Yeah the thumbor system is in 2.7.

peterdemin commented:
Hey guys, I didn't get your opinion on using `pip-compile-multi verify` inside `tools/test-locked-requirements`.
I explained changes in logic in a comment for https://github.com/zulip/zulip/pull/8079/commits/f018748f5e20e890f84a8e280a99493658235a76
Let's make a decision.
cc @rht @adnrs96 @timabbott @gnprice

timabbott commented:
@peterdemin I'm super busy right now.  Can you provide a bit more of a summary for (1) In what ways `pip-compile-multi` is better/different than the old system (other than getting the delete code, which is nice) and (2) why you need so many changes to our `requirements` files?  E.g. why the conversion to use `~=` ?  

As a sidenote, it's kinda not great that you have so many changes to the versions of individual packages mixed into this PR; ideally, the tooling change would be independent from the packages we end up with.  But maybe that's tied in with the `~=` switch?

peterdemin commented:
Right, this must be explained.
1) pip-compile-multi is a universal tool, that automatically discovers which requirements files need to be compiled (locked), thus it needs little to no configuration. It allows keeping references between requirements files, so that resulting files are smaller and don't have duplicates. It evolves and provides abilities that zulip might need in future, like adding packages hashes. `pip-compile-multi verify` runs faster (instantly) than old `tools/test-locked-requirements` because it doesn't recompile all `.txt` files, but just checks `.in` files hashes.
2) There are a lot of changes to requirements files because `pip-compile-multi` slightly changes a way of working with requirements. First - `common.in` is compiled to `common.txt` and then referenced from other files. Second - `.in` contain only direct dependencies, packages are unconstrained or constrained to avoid known incompatibilities, while packages in `.txt` files are hard-pinned. Using compatible release (`~=`) is a temporary measure to avoid major dependency updates while changing toolset.
Actually, I don't see much use in hard-pinning packages in `.in` files - it makes it harder to update them.

`pip-compile-multi` allows adding new workflow - periodic dependency update. Say once a month you create a branch and run `pip-compile-multi` there and check that CI tests are passing (and merge). That's the easiest way of not having to deal with backward incompatible releases while using the freshest packages with all the bug fixes and performance improvements.

I can separate this PR into two:
1) Change the toolset.
2) Execute new tools to update dependencies.
But I think it is better this way since you can see the change and the effect (and CI tests pass).

timabbott commented:
The story for using `pip-compile-multi` makes sense.

> Actually, I don't see much use in hard-pinning packages in .in files - it makes it harder to update them.

This is important to provide predictability; with a project of Zulip's size, dependencies are constantly releasing breaking upgrades, and we need to manage that explicitly.  So our policy is to pin explicit versions everywhere.  

I think if you just stuck with the hard-pinned versions in your PR, you'd be able to avoid merge conflicts (or material changes to the `.in` files), and that'd make it much easier to maintain this branch and eventually integrate it.  We can always switch from doing hard-pinning later; what's important for now is to ensure that we're not breaking anything, and having a bunch of effectively unrelated package version changes mixed in with the tooling change is a recipe for a PR that gets stuck.

Does that make sense to you @peterdemin?

peterdemin commented:
Yeah, I hoped to make this in bulk, relying on the test suite. Ok, I'll revert to hard-pins in `.in` files.
However, I don't quite understand the argument on predictability.
Packages are installed from `.txt` files, that have hard pins, so the builds are predictable.
I'm a proponent of frequent updates - once a problematic dependency version is encountered, it should be constrained in `.in` files (along with the creation of an Issue for adopting new version).

timabbott commented:
> Packages are installed from .txt files, that have hard pins, so the builds are predictable.

When we need to update one package (or add a new dep), we don't want updating the rest of our dependencies to be part of that PR.  So, instead, at least once a month, we do a big manual update like #8371.

peterdemin commented:
I see. I want to automate this case with `pip-compile-multi`.

One way to solve this is to have `--direct-only` flag, that will leave only direct dependencies in the generated file.
Then zulip will have 3 layers of requirements files:
1) `.unc` - Unconstrained list of package names, except for known compatibility problems.
2) `.prim` - Primary hard-pinned dependencies - that's what currently stored in `.in` files.
3) `.txt` - a full list of hard-pinned dependencies.

For adding new dependency during feature PR it is added both to `.unc` and to `.prim` file and then a full list of dependencies is compiled in `.txt` files.

For the regular update of primary dependencies `.unc` files are compiled to `.prim` with `--direct-only` flag, `.prim` files are compiled to `.txt`.

I think it will make bulk dependency updates (like in #8371) easier to do.

peterdemin commented:
Another option is to run `pip-compile-multi` with `--no-upgrade`, that will leave all satisfying dependencies as is.

peterdemin commented:
To make this PR easier to review I removed *all* version updates.
Leaving only essential pieces:
1) Update tooling;
2) Add `common.txt` (with all common packages locked) that is referenced from other .txt files.

peterdemin commented:
I added `--no-upgrade` option to `pip-compile-multi` and rebased the branch. So now it fully supports zulip's use case.
@timabbott, what do you think?

peterdemin commented:
Hey @timabbott, do you have the time to review the PR and merge/close it?
If so, I'll resolve the conflicts.

timabbott commented:
Closing, since we're currently looking at direction of using `nix` to make our production environment even more portable, so we're unlikely to want to do further changes.

Thanks for all your work on this @peterdemin! I think `pip-compile-multi` is a cool project.

peterdemin commented:
Sounds good. I'm curious, how does nix address this problem?

rht commented:
Regarding with portability, you can install `nix` on any *NIX systems, and the same `nix` expressions are used in all those systems. Code branching to address various distributions, as you can see in https://github.com/zulip/zulip/blob/5ee4f717019aaceb5408de807e351165bc3da497/puppet/zulip/manifests/common.pp#L3-L39 and lots of other places, is no longer required.
